### PR TITLE
Use cached data to validate policy admissions

### PR DESCRIPTION
**Pull request by olix0r**
Currently, the policy-controller's admission server issues a request to
Kubernetes every admission review request. This is unnecessary, since
we already maintain a cache of the relevant data structure in the
process.

This change updates the admission server to hold a reference to the
process's `Index` data structure. Each time an admission request is
received, the admission server accesses the index to lookup servers that
already exist in the cluster.

Note that it's possible that the index may not be 100% up to date, as
event delivery can be delayed. I think it's okay to make the tradeoff
that the admission controller is best-effort in favor of reducing
resource usage.

Signed-off-by: Oliver Gould <ver@buoyant.io>

### CODE REVIEW

File: policy-controller/k8s/api/src/lib.rs [hunk 1]
```diff
@@ -5,38 +5,9 @@ pub mod labels;
...
-    E: Stream<Item = watcher::Event<T>>,
-    F: Fn(&mut S, watcher::Event<T>),
-{
-    tokio::pin!(events);
-    while let Some(ev) = events.next().await {
-        process(&mut *store.lock(), ev);
-    }
-
-    tracing::warn!("k8s event stream terminated");
-}
```
Comment (olix0r) [inline policy-controller/k8s/api/src/lib.rs:42]: Should have been removed in d4543cd86e427b241ce961b50dd83b1738c0b069

**Review by kleimkuhler — APPROVED** [review]
Cool this makes sense 👍

**Review by adleong — APPROVED** [review]
Do we know how long it takes for the index to be updated?  E.g. if we reject a Server creation because of a conflict that has actually been deleted but still remains in the index, can the user simply attempt the creation again and expect it to succeed on the second try?

**Review by alpeb — APPROVED** [review]
LGTM :+1:

**Comment by olix0r** [comment]
> Do we know how long it takes for the index to be updated? E.g. if we reject a Server creation because of a conflict that has actually been deleted but still remains in the index, can the user simply attempt the creation again and expect it to succeed on the second try?

There's what we can expect and then there's what we can guarantee... We expect the index to basically be up-to-date, so yes a retry should generally be fine. I think we can only guarantee that we're up to date with 290s (the max request timeout), after which we'll restart the watch and be sure to get updated data. We could reduce that timeout if we saw consistency issues...

**Pull request merged**